aarch64: Reorganize `test_create_device`

TimePrinciple opened the pull request:
### Summary of the PR

Add `has_device_attr` check in both `set_supported_nr_irqs` and `request_gic_init`, effectively enabling reuse of `request_gic_init` in `test_create_device`.

I was going through these tests while introducing RISC-V ioctls. Some minor improvements :)

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR have Signed-Off-By trailers (with
  `git commit -s`), and the commit message has max 60 characters for the
  summary and max 75 characters for each description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

RuoqingHe committed:
aarch64: Reorganize `test_create_device`

Add `has_device_attr` check in both `set_supported_nr_irqs` and
`request_gic_init`, effectively enabling reuse of `request_gic_init` in
`test_create_device`.

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

roypat commented on src/ioctls/device.rs:
```diff
@@ -292,15 +289,7 @@ mod tests {
         set_supported_nr_irqs(&device_fd, 128);
 
         // Following attribute works with VGIC, they should be accepted.
```
Let's also update this comment (something like "initialize valid vgic"?)

TimePrinciple replied:
Updated :)

roypat commented on src/ioctls/vm.rs:
```diff
@@ -1942,6 +1943,7 @@ pub(crate) fn request_gic_init(vgic: &DeviceFd) {
         addr: 0,
         flags: 0,
     };
+    assert!(vgic.has_device_attr(&vgic_attr).is_ok());
```
Kinda unrelated to your PR, but could we do
```suggestion
    vgic.has_device_attr(&vgic_attr).unwrap();
```
instead? `assert((...).is_ok())` gives very unhelpful error messages, because it doesn't print out the `Err` variant (should an error happen).

TimePrinciple replied:
A very good point, should I replace these `assert!`s with `unwrap()` within the lines this PR touched, or do it globally 🤔

roypat replied:
If you want to, I'd really appreciate a global change, although I don't insist on it, because it can be quite tedious (for reference, the PR when I did it on Firecracker: https://github.com/firecracker-microvm/firecracker/pull/4321).

So just changing it for the function you touch here is fine!

RuoqingHe committed:
chore: Eliminate use of `assert!((...).is_ok())`

As @roypat pointed out, and quote:

"Asserting on .is_ok()/.is_err() leads to hard to debug failures (as if
the test fails, it will only say "assertion failed: false". We replace
these with `.unwrap()`, which also prints the exact error variant that
was unexpectedly encountered (we can to this these days thanks to
efforts to implement Display and Debug for our error types). If the
assert!((...).is_ok()) was followed by an .unwrap() anyway, we just drop
the assert."

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

RuoqingHe committed:
chore: Enable clippy::assertions_on_result_states

As @roypat pointed out, and quote:

"This lint disallows asserttions on is_ok()/is_err() in favor of either
using unwrap (so that at least if the test fails, we the failure message
will contain the actual failure reason instead of just "was not
ok/err"), or actually matching the specific variant."

Signed-off-by: Ruoqing He <heruoqing@iscas.ac.cn>

roypat commented on .cargo/config.toml:
```diff
@@ -1,3 +1,8 @@
+[build]
+rustflags = [
+  "-Wclippy::assertions_on_result_states",
+]
```
This should be a `#![warn(clippy::assertions_on_result_states)]` in `lib.rs` :)

We did it like this in Firecracker for unrelated reasons that don't apply to rust-vmm

TimePrinciple replied:
> This should be a `#![warn(clippy::assertions_on_result_states)]` in `lib.rs` :)
> 
> We did it like this in Firecracker for unrelated reasons that don't apply to rust-vmm

Noted

rbradford reviewed (approved):
Great improvements. I especially like the assertion cleanups.

TimePrinciple commented:
> Great improvements. I especially like the assertion cleanups.

If you like, I will do it in cloud-hypervisor :)
